### PR TITLE
Oracle: Economic model (Burning SysFee)

### PR DESCRIPTION
**Pull request by shargon**
- SysFee of oracles (request and response) goes to Oracle multisig if it has less than X otherwise it will be distributed for oracle nodes.
- User will pay for the response extra cost.
- User will pay for the filter cost.

### CODE REVIEW

**Comment by shargon** [comment]
I'm concerned about lazy oracles, they can't copy the results, but they can do nothing because the signature is not required to be fully signed, maybe the multisig should be more strict (only one less)

File: src/neo/SmartContract/Native/Oracle/OracleContract.cs [hunk 1]
```diff
@@ -47,6 +47,30 @@ internal override bool Initialize(ApplicationEngine engine)
...
+        /// Distribute reward
+        /// </summary>
+        [ContractMethod(0_03000000, ContractParameterType.Integer)]
+        private StackItem DistributeRewards(ApplicationEngine engine, Array args)
+        {
+            var account = GetOracleMultiSigAddress(engine.Snapshot);
+            var validators = GetOracleValidators(engine.Snapshot);
+            var balance = GAS.BalanceOf(engine.Snapshot, account);
+            var amount = balance / validators.Length;
+
```
Comment (Tommo-L) [inline src/neo/SmartContract/Native/Oracle/OracleContract.cs:60]: I think we should consider the case of validators changed.

File: src/neo/SmartContract/Native/Tokens/GasToken.cs [hunk 2]
```diff
@@ -32,11 +32,40 @@ internal override bool Initialize(ApplicationEngine engine)
...
+
             ECPoint[] validators = NEO.GetNextBlockValidators(engine.Snapshot);
             UInt160 primary = Contract.CreateSignatureRedeemScript(validators[engine.Snapshot.PersistingBlock.ConsensusData.PrimaryIndex]).ToScriptHash();
-            Mint(engine, primary, engine.Snapshot.PersistingBlock.Transactions.Sum(p => p.NetworkFee));
+            Mint(engine, primary, networkFee);
+
+            if (oracleFee > 0)
+            {
+                primary = Oracle.GetOracleMultiSigAddress(engine.Snapshot);
+                Mint(engine, primary, oracleFee);
```
Comment (Tommo-L) [inline src/neo/SmartContract/Native/Tokens/GasToken.cs:67]: I think we can distribute rewards to the oracle validators directly here, or move this logic into OracleContract.

Comment (shargon) [reply src/neo/SmartContract/Native/Tokens/GasToken.cs:67]: Yes, maybe here we distribute exactly to this OracleNodes.

**Comment by shargon** [comment]
@Tommo-L It was changed, but now, Oracle nodes need to send Gas to OracleMultiSignature in order to be able to send the response Tx. Maybe if the amount of this account is less than X we should move all to this address.

**Comment by erikzhang** [comment]
We should decide the model first and then develop it. In addition, I think we should finally consider the economic model, and we can design other parts first.

**Comment by shargon** [comment]
> We should decide the model first and then develop it. In addition, I think we should finally consider the economic model, and we can design other parts first.

I think that this is the last part, the rest of the oracle could be optimized and should be tested, but it's working now. We can use https://github.com/neo-project/neo/issues/1279 for taking the decision

File: src/neo/Wallets/Wallet.cs [hunk 3]
```diff
@@ -345,6 +345,10 @@ private Transaction MakeTransaction(StoreView snapshot, byte[] script, Transacti
 
                     if (oracleRequests?.Count > 0)
                     {
+                        // We should pay for the response extra cost
+
+                        tx.SystemFee += (oracleCache.Size * NativeContract.Policy.GetFeePerByte(snapshot));
+
```
Comment (Tommo-L) [inline src/neo/Wallets/Wallet.cs:351]: We can use oracle basic fee to repalce this.

```js
OracleFee = oracle.basicFee + oracle.Size * NativeContract.Policy.GetFeePerByte(snapshot)
```

File: src/neo/SmartContract/Native/Tokens/GasToken.cs [hunk 4]
```diff
@@ -32,11 +32,55 @@ internal override bool Initialize(ApplicationEngine engine)
...
+                }
+
+                // Distribute oracle fee
+
+                validators = Oracle.GetOracleValidators(engine.Snapshot);
+                var amount = oracleFee / validators.Length;
+
+                foreach (var validator in validators)
+                {
+                    // Transfer check the oracle's multi signature
```
Comment (Tommo-L) [inline src/neo/SmartContract/Native/Tokens/GasToken.cs:79]: Will it be better if we could move this incentive model to `OracleContract.OnPersist` method?

Comment (shargon) [reply src/neo/SmartContract/Native/Tokens/GasToken.cs:79]: Then Gas will burn all of them first

Comment (Tommo-L) [reply src/neo/SmartContract/Native/Tokens/GasToken.cs:79]: Maybe it's okay to put here, as it's  more efficient.

**Pull request closed without merging**